### ISSUE TITLE
M5: Advertise Policy Templates in ServiceAccessInformation

### ISSUE THREAD

**Issue by davidjwbbc** [5G-MAG/rt-5gms-application-function#91]
# Description

In order for a UE to use Dynamic Policies, while using the 5G Media Streaming services, it needs to be told which dynamic policies are available to it. The list of dynamic policies are conveyed via the `ServiceAccessInformation` message for the `retrieveServiceAccessInformation` API method on the interface at reference point M5.

When a UE encounters a media manifest that indicates dynamic policy identifiers for representations it can then request the that the correct policy is in place when changing to the marked up representation. It may also ignore representations whose policy is not yet available.

# Design

## ServiceAccessInformation changes

When the _ServiceAccessInformation_ resource is generated for a Provisioning Session and the Provisioning Session has one or more Policy Templates defined which are in the *READY* state, then a `dynamicPolicyInvocationConfiguration` (`OpenAPI_service_access_information_resource_dynamic_policy_invocation_configuration_t`) is set. This is created to contain:

- A *server_addresses* `OpenAPI_list_t` which has one entry containing the URL of the M5 interface of the AF used during the M5 request for the _ServiceAccessInformation_ (the listening address of the M5 interface cannot be used as this may be configured to be the "catch all" address, "`0.0.0.0`" for IPv4 or "`::`" for IPv6).
  - The URL in the list is `"{request.protocol}://{request.header("Host")}/3gpp-m5/v2"` to represent the `apiRoot` of the AF and the M5 protocol version accepted.
- The *valid_policy_template_ids* field contains the list of all policy template identifiers for those Policy Templates that are in the *READY* state in the Provisioning Session.
- The *sdf_methods* field is fixed as a list of strings containing one entry with `5_TUPLE` in it.
- The *external_references* field contains the set of *externalReference* values from the Policy Templates in the Provisioning Session that are in the *READY* state.

The `ServiceAccessInformation` caches for the Provisioning Session are invalidated whenever a Policy Template for the Provisioning Session transitions to or from the *READY* state because any such state change alters the list of valid Policy Template identifiers.

# Relevant Specifications

TS 26.512 (v17.6.0)
- Clauses 4.7.3 & 11.2
- Annex C.4.1

**Comment by devbbc** [comment]
Started implementing the internal data structures and  obtaining the  `valid_policy_template_ids` field from the  policy templates in the ready state associated with a Provisioning session.

**Comment by devbbc** [comment]
Retrieving the `serviceAccessInformation` cab now advertise the valid policy templates.

Sample output:

```
$ curl -H 'User-Agent: AF' -v 'http://127.0.0.24:7777/3gpp-m5/v2/service-access-information/b54a1428-6848-41ee-aa9a-ad12c2bbb05a'

> GET /3gpp-m5/v2/service-access-information/b54a1428-6848-41ee-aa9a-ad12c2bbb05a HTTP/1.1
> Host: 127.0.0.24:7777
> Accept: */*
> User-Agent: AF
> 

< HTTP/1.1 200 OK
< Connection: Keep-Alive
< Content-Length: 486
< Server: 5GMSAF-af.localdomain/17 (info.title=M5_ServiceAccessInformation; info.version=2.3.0) rt-5gms-application-function/1.3.0
< Cache-Control: max-age=60
< Last-Modified: Wed, 11 Oct 2023 15:14:20 GMT
< ETag: a00f21994c4fd460afc7d973a6ceb75f03a8c20b5d814400cb28dffdece4df9d
< Content-Type: application/json
< Date: Wed, 11 Oct 2023 15:14:19 GMT
< 
{
	"provisioningSessionId":	"b54a1428-6848-41ee-aa9a-ad12c2bbb05a",
	"provisioningSessionType":	"DOWNLINK",
	"streamingAccess":	{
	},
	"dynamicPolicyInvocationConfiguration":	{
		"serverAddresses":	["http://127.0.0.24:7777/3gpp-m5/v2/"],
		"validPolicyTemplateIds":	["c291c720-6848-41ee-aa9a-ad12c2bbb05a", "ce073c84-6848-41ee-aa9a-ad12c2bbb05a"],
		"sdfMethods":	["5_TUPLE"],
		"externalReferences":	["9a2d2f3c-5158-4ec6-8989-cfb03c0ecf5d", "8b3e2f3c-5158-4ec6-8989-cfb03c0ece4c"]
	}
}
```

**Comment by rjb1000** [comment]
Looks good. Just to note that the _externalReferences_ are not expected to be UUIDs in reality. They are more likely to be short strings describing Service Operation Points, such as "HD" or "10M". They are intended to match values in the DASH MPD.

**Comment by rjb1000** [comment]
Question: how are you populating _sdfMethods_?

I can't see anything from the M1 _PolicyTemplate_ to drive this, so am I right that it comes from local system configuration?

**Comment by devbbc** [comment]
> how are you populating sdfMethods?

5-Tuple is the only Service Date Flow Method that the implementation supports. So, the AF is a returning a fixed value `5_TUPLE`.

In the future, this could be a local system configuration.

**Comment by devbbc** [comment]
> The ServiceAccessInformation caches for the Provisioning Session are invalidated whenever a Policy Template for the Provisioning Session transitions to or from the READY state because any such state change alters the list of valid Policy Template identifiers.

This Service Access Information cache invalidation part is implemented as part of #70.

**Comment by davidjwbbc** [comment]
Tested the M5 Service Access Information interface for when policy templates are added or removed.

Found that the SAI was reporting the dynamic policies structure with no template ids listed when the policy template is deleted. I've adjusted the code to omit the dynamic policies structure when there are no active policy templates available for the provisioning session.

Also tested the behaviour with multiple policy templates and found a memory use after free issue when an earlier listed policy template is deleted. Fixed the use after free by comparing the structure pointers rather than the id strings held within and aborting the find-and-delete loop when a match is found and deleted.

This interface now works as expected. I'll push the fixes in a new commit before doing the PR (along with a commit for the python tools updates for issue #106). I'll move this issue to PR pending when I've committed the fixes.
